### PR TITLE
xds: WRR scheduler clips weights

### PR DESCRIPTION
**Pull request by tonyjongyoonan**
In production, the static stride scheduler in Stubby was [failing](https://b.corp.google.com/issues/276292666) in rare cases where a subchannel's weight was extremely close to 0. In cases like these, these subchannels would never be selected. Or even worse, the scheduler would never make a pick. Stubby chose to clip weights to fix this issue.

While this doesn't seem to be a problem for gRPC, it may be good for gRPC to also clip weights to optimize for a worst case possibility where a large difference in subchannel weights causes us to frequently traverse through the entire subchannel list before making a pick. By clipping weights, we decrease the number of cycles it takes to make a pick in these worst case possibilities. While more RPCs will be sent to poorly performing subchannels, this is not necessarily a downside. We can clip weights even more if we want to save even more cycles.

Since we're taking this opportunity to optimize for this possibility, we will also fix some flakiness in WRR tests caused by the scheduler no longer being deterministic unlike the previous EDFScheduler, as described in #10479. Verbose test cases were also removed. 

It might be good to make a note of this edge case somewhere so that other teams don't run into the same issue when implementing, especially since [go/static-stride-scheduler](https://g3doc.corp.google.com/production/rpc/stubs/g3doc/loadbalancing/static_stride_scheduler.md?cl=head) isn't updated to make note of this.

### CODE REVIEW

**Comment by tonyjongyoonan** [comment]
Two things: weights calculated in the creation of the scheduler were changed from `short` to `int`, because `short` reads negative numbers unless `Short.toUnsignedInt()` is called, which would mess up calculations. Adding the cast to an `uint16` would make the code a little bit less legible, but if the performance difference is worth the legibility, this can quickly be changed.

Additionally, the `pickByWeight_largeWeight` test cases were replaced with unit tests for the schedulers, as these weights are capped and there isn't really a good way to test this now. If it is still desirable to test these beyond the unit tests (I am assuming this is the case), some test cases with predictable weights can be created.

File: xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java [hunk 1]
```diff
@@ -350,36 +350,91 @@ static final class StaticStrideScheduler {
...
+    // M*kMaxRatio).
+    //
+    // This is done to limit the number of rounds for picks.
+    private static final double K_MAX_RATIO = 10;
+
+    // Assuming the mean of all known weights is M, StaticStrideScheduler will cap
+    // from below all known weights to M*kMinRatio.
+    //
+    // This is done as a performance optimization for edge cases when channels with
+    // large weights are non-accepting (and thus WeightedRoundRobin will retry
```
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:364]: This might not impact gRPC. The list passed into this scheduler has already filtered out the non-accepting subchannels (and the definition of non-accepting is not quite the same). That means the worst-case time is O(n), because at the very least we'd pick the subchannel with the max weight.

It is possible in gRPC for a transport not to accept an RPC, but in that case the RPC is queued and we wait for the LB to notice the subchannel is no longer available. That will then generate a new scheduler with a new max weight.

Comment (tonyjongyoonan) [reply xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:364]: Aha, so I guess we wouldn't run into the non-accepting, infinite loop issue that Stubby did, but this is still useful to optimize for cases where there's a great difference in weights?

Comment (ejona86) [reply xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:364]: I think it is useful to put in place a bound.

File: xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java [hunk 2]
```diff
@@ -351,35 +351,83 @@ static final class StaticStrideScheduler {
...
+      // Note that, since we cap the weights to stay within K_MAX_RATIO, `meanWeight` might not
+      // match the actual mean of the values that end up in the scheduler.
+      double scalingFactor = K_MAX_WEIGHT / unscaledMaxWeight;
+      int meanWeight = (int) Math.round(scalingFactor * unscaledMeanWeight);
+
+      // We compute `weightLowerBound` and cap it to 1 from below so that in the
+      // worst case we represent tiny weights as 1 but not as 0 (which would cause
+      // an infinite loop as in b/276292666). This capping to 1 is probably only
+      // useful in case someone misconfigures kMinRatio to be very small.
+      int weightLowerBound = (int) Math.max(1, Math.round(K_MIN_RATIO * meanWeight));
```
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:422]: I don't feel entirely comfortable doing this conversion in integer space, as the others have been using floats. I'd suggest:
```java
int weightLowerBound = (int) Math.ceil(scalingFactor * unscaledMeanWeight * K_MIN_RATIO);
```

Note that using ceil gets rid of the "rounds to 0" problem.

File: xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java [hunk 3]
```diff
@@ -351,35 +351,83 @@ static final class StaticStrideScheduler {
...
       short[] scaledWeights = new short[numChannels];
       for (int i = 0; i < numChannels; i++) {
         if (weights[i] <= 0) {
-          scaledWeights[i] = meanWeight;
+          scaledWeights[i] = (short) meanWeight;
         } else {
-          scaledWeights[i] = (short) Math.round(weights[i] * scalingFactor);
+          double weightUpperBound = Math.min(weights[i], unscaledMaxWeight);
```
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:428]: Oh, boy, is that a name asking to be misunderstood. I don't have a good alternative name. Although `doubleWeight` is probably better.

**Review by ejona86 — COMMENTED** [review]
I'm pretty disappointed we have to have the three-stage min-scale-max handling at the end of the conversion. But I do see why. I don't think we can avoid it, but it makes it more bug-prone and harder to understand.

File: xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java [hunk 4]
```diff
@@ -351,35 +351,83 @@ static final class StaticStrideScheduler {
...
+    //
+    // The current values of 10 and 0.1 were chosen without any experimenting. It should
+    // decrease the amount of sequences that the scheduler must traverse through in order
+    // to pick a high weight subchannel in such corner cases.
+    // But, it also makes WeightedRoundRobin to send slightly more requests to
+    // potentially very bad tasks (that would have near-zero weights) than zero.
+    // This is not necessarily a downside, though. Perhaps this is not a problem at
+    // all, and we can increase this value if needed to save CPU cycles.
+    //
+    // Note that this class treats weights that are exactly equal to zero as unknown
```
[the remark this answers is not in the document]
Comment (tonyjongyoonan) [reply xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:371]: Currently, negative/zero weights will be set to `mean` while weights that are close to zero (e.g. 0.000001), can be set to `mean*kMinRatio`. The EDFScheduler had a minimum weight of 0.0001, which we got rid of with the new scheduler. I'm getting confused on what's supposed to be "right" since we seem to be deviating from WRR with this weight clip.

But yes, I see what you mean about making this a TODO. What do you think the desired behavior here should be?

[the remark this answers is not in the document]
Comment (ejona86) [reply xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:370]: It is normal good practice to not compare floats for equality directly but within a range of a small "epsilon", because rounding and accumulating error causes havoc. But we are comparing to literal zeros that are in the code; zero isn't the result of a computation. That makes it "work" to use equality checks, although not everyone would agree that it is "okay."

I was arguing this comment doesn't belong here. This is about the ratios which has nothing to do with "zero means use average." The code doesn't confuse the two cases. I'd be fine with this paragraph being deleted.

**Review by ejona86 — APPROVED** [review]

**Comment by tonyjongyoonan** [comment]
@ejona86 I changed the unscaledMaxWeight to be set to 1 in the case where we have all invalid weights and need to fall back to round robin--feel free to take a quick look if you have the time?

Also, [TSAN](https://fusion2.corp.google.com/invocations/5953480b-ff61-47d5-9733-66f2f6ca9afe/targets/%2F%2Fthird_party%2Fjava_src%2Fgrpc%2Fxds:src%2Ftest%2Fjava%2Fio%2Fgrpc%2Fxds%2FWeightedRoundRobinLoadBalancerTest/tests) seems to be slow for `testAllInvalidWeightsIsRoundRobin`, which is a test case that only makes six picks. I did check that we make the first pick inside pick, but is this normal? `rrFallback` tests the same thing but moves quicker even with a thousand picks.

File: xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java [hunk 5]
```diff
@@ -462,7 +459,10 @@ long getSequence() {
      * an offset that varies per backend index is also included to the calculation.
      */
     int pick() {
+      int i = 0;
```
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:462]: Clean up.

File: xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java [hunk 6]
```diff
@@ -390,22 +384,25 @@ static final class StaticStrideScheduler {
...
       // Scales weights s.t. max(weights) == K_MAX_WEIGHT, meanWeight is scaled accordingly.
       // Note that, since we cap the weights to stay within K_MAX_RATIO, `meanWeight` might not
       // match the actual mean of the values that end up in the scheduler.
       double scalingFactor = K_MAX_WEIGHT / unscaledMaxWeight;
       int meanWeight = (int) Math.round(scalingFactor * unscaledMeanWeight);
-
       // We compute `weightLowerBound` and clamp it to 1 from below so that in the
       // worst case, we represent tiny weights as 1.
-      int weightLowerBound = (int) Math.ceil(scalingFactor * unscaledMeanWeight * K_MIN_RATIO);
+      int weightLowerBound = (int) Math.ceil((scalingFactor * unscaledMeanWeight * K_MIN_RATIO));
```
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:405]: Remove extraneous parens?

File: xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java [hunk 7]
```diff
@@ -390,22 +384,25 @@ static final class StaticStrideScheduler {
...
-      double unscaledMeanWeight = numWeightedChannels > 0
-          ? sumWeight / numWeightedChannels : 1;
-
       // Adjust max value s.t. ratio does not exceed K_MAX_RATIO. This should
       // ensure that we on average do at most K_MAX_RATIO rounds for picks.
-      unscaledMaxWeight = Math.min(unscaledMaxWeight, (float) (K_MAX_RATIO * unscaledMeanWeight));
+      if (numWeightedChannels > 0) {
+        unscaledMeanWeight = sumWeight / numWeightedChannels;
+        unscaledMaxWeight = Math.min(unscaledMaxWeight, (float) (K_MAX_RATIO * unscaledMeanWeight));
+      // Fall back to round robin if all values are non-positives
```
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:392]: Put this on the same line as the else (`} else { // Fall back...`) or first line within the else block.

File: xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java (hunk 5 above)
Comment (ejona86) [reply xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java:461]: Idea for checking this during tests: take the difference of AtomicInteger before/after calling this method. Probably best not to do that in this PR.

**Review by ejona86 — APPROVED** [review]
